bigtable/metrics.go

```diff
@@ -0,0 +1,487 @@
+	}
+
+	if emulatorAddr := os.Getenv("BIGTABLE_EMULATOR_HOST"); emulatorAddr != "" {
+		// Do not emit metrics when emulator is being used
+		return tracerFactory, nil
+	}
+
+	var meterProvider *sdkmetric.MeterProvider
+	if metricsProvider == nil {
+		// Create default meter provider
+		mpOptions, err := getBuiltInMeterProviderOptions(ctx, project)
+		if err != nil {
+			return tracerFactory, err
+		}
+		meterProvider = sdkmetric.NewMeterProvider(mpOptions)
+
+		tracerFactory.builtinEnabled = true
+		tracerFactory.shutdown = func() { meterProvider.Shutdown(ctx) }
+	} else {
+		switch v := metricsProvider.(type) {
+		case CustomOpenTelemetryMetricsProvider:
+			// User provided meter provider
+			tracerFactory.builtinEnabled = true
+			meterProvider = v.MeterProvider
+		case NoopMetricsProvider:
+			tracerFactory.builtinEnabled = false
+			return tracerFactory, nil
+		default:
+			tracerFactory.builtinEnabled = false
+			return tracerFactory, errors.New("Unknown MetricsProvider type")
+		}
+	}
+
+	// Create meter and instruments
+	meter := meterProvider.Meter(builtInMetricsMeterName, metric.WithInstrumentationVersion(internal.Version))
+	err = tracerFactory.createInstruments(meter)
+	return tracerFactory, err
+}
+
+func (tf *builtinMetricsTracerFactory) createInstruments(meter metric.Meter) error {
+	var err error
+
+	// Create operation_latencies
+	tf.operationLatencies, err = meter.Float64Histogram(
+		metricNameOperationLatencies,
+		metric.WithDescription("Total time until final operation success or failure, including retries and backoff."),
+		metric.WithUnit("ms"),
+		metric.WithExplicitBucketBoundaries(bucketBounds...),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create attempt_latencies
+	tf.attemptLatencies, err = meter.Float64Histogram(
+		metricNameAttemptLatencies,
+		metric.WithDescription("Client observed latency per RPC attempt."),
+		metric.WithUnit("ms"),
+		metric.WithExplicitBucketBoundaries(bucketBounds...),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create server_latencies
+	tf.serverLatencies, err = meter.Float64Histogram(
+		metricNameServerLatencies,
+		metric.WithDescription("The latency measured from the moment that the RPC entered the Google data center until the RPC was completed."),
+		metric.WithUnit("ms"),
+		metric.WithExplicitBucketBoundaries(bucketBounds...),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create retry_count
+	tf.retryCount, err = meter.Int64Counter(
+		metricNameRetryCount,
+		metric.WithDescription("The number of additional RPCs sent after the initial attempt."),
+	)
+	return err
+}
+
+// builtinMetricsTracer is created one per function call
+// It is used to store metric instruments, attribute values
+// and other data required to obtain and record them
+type builtinMetricsTracer struct {
+	ctx            context.Context
+	builtInEnabled bool
+	opStartTime    time.Time
+
+	// attributes that are specific to a client instance and
+	// do not change across different function calls on client
```

Comment:
does function call mean operation or attempt here?

Reply:
Also consider storing all of the client scoped values together in a struct

Reply:
> does function call mean operation or attempt here?

Updated the comment